cmd/composer-cli/blueprints/diff.go

```diff
@@ -25,7 +31,189 @@ func init() {
+	if err != nil {
+		return root.ExecutionError(cmd, err.Error())
+	}
+
+	// Everything after args[2] is passed to diff
+	// default to --color and -u if nothing passed
+	diffArgs := args[3:]
+	if len(diffArgs) == 0 {
+		diffArgs = []string{"--color", "-u"}
+	}
+
+	err = runDiff(fromBlueprint, args[1], toBlueprint, args[2], diffArgs)
+	if err != nil {
+		return root.ExecutionError(cmd, "%s", err)
+	}
+
+	return nil
+}
+
+func getBlueprint(name, commit string) (string, error) {
+	if commit == "WORKSPACE" {
+		// If nothing has been pushed to the /blueprints/workspace this will be the latest
+		bps, resp, err := root.Client.GetBlueprintsTOML([]string{name})
+		if err != nil {
+			return "", err
+		}
+		if resp != nil && !resp.Status {
+			return "", fmt.Errorf(strings.Join(resp.AllErrors(), ", "))
+		}
+		if len(bps) == 0 {
+			return "", fmt.Errorf("no blueprints")
+		}
+		return bps[0], nil
+	} else if commit == "NEWEST" {
+		// Get the list of changes for this blueprint
+		bps, resp, err := root.Client.GetBlueprintsChanges([]string{name})
+		if err != nil {
+			return "", err
+		}
+		if len(resp) > 0 {
+			return "", fmt.Errorf(resp[0].String())
+		}
+		if len(bps) < 1 {
+			return "", fmt.Errorf("no blueprints")
+		}
+		if len(bps[0].Changes) < 1 {
+			return "", fmt.Errorf("no NEWEST commit for %s", name)
+		}
+		commit = bps[0].Changes[0].Commit
+	}
+
+	// Return the blueprint commit
+	bp, resp, err := root.Client.GetBlueprintChangeTOML(name, commit)
+	if err != nil {
+		return "", err
+	}
+	if resp != nil && !resp.Status {
+		return "", fmt.Errorf(strings.Join(resp.AllErrors(), ", "))
+	}
+	return bp, nil
+}
 
+// runDiff runs the system diff utility from a temporary directory
+func runDiff(fromBlueprint, fromCommit, toBlueprint, toCommit string, diffArgs []string) error {
+	cwd, err := os.Getwd()
+	if err != nil {
+		return err
+	}
+
+	// Make a temporary directory to save the blueprints into
+	// defer cleanup so it doesn't ever get left behind
+	tmpDir, err := os.MkdirTemp("", "bp-diff-*")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(tmpDir)
+	err = os.Chdir(tmpDir)
+	if err != nil {
+		return err
+	}
+	defer os.Chdir(cwd) //nolint:errcheck
+
+	// Use saveBlueprint from save.go
+	fromFilename, err := saveBlueprint(fromBlueprint, fromCommit, "")
+	if err != nil {
+		return err
+	}
+	diffArgs = append(diffArgs, fromFilename)
+
+	toFilename, err := saveBlueprint(toBlueprint, toCommit, "")
+	if err != nil {
+		return err
+	}
+	diffArgs = append(diffArgs, toFilename)
+
+	_, err = exec.LookPath("diff")
+	if err != nil {
+		return fmt.Errorf("The diff utility is required, please install it")
+	}
+
+	cmd := exec.Command("diff", diffArgs...)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	err = cmd.Start()
+	if err != nil {
+		return fmt.Errorf("diff error: %s", err)
+	}
+	err = cmd.Wait()
+	if err != nil {
+		// 0 == same file 1 == different file 2 == errors
+		if cmd.ProcessState.ExitCode() > 1 {
```

Comment:
I wonder if we should check if `err` is `ExitError` and then access the process state through the error. Feels slightly cleaner to me.

Also, we currently don't handle any other errors than `ExitError`, which might be worth doing.